### ISSUE TITLE
Load tests: Explore possibilities of JMeter for QA tests

### ISSUE THREAD

**Issue by hofmannc** [hofmannc/MPDL-AutomatedTesting#5]
Goal: 
Find out if Jmeter in combi with Selenium and Jenkins can be used for the implementation of loadtests as expected.
--> Insights will be also shared with team by a meeting (to be planned).
--> will be updated to concept #3 "Create concept "Lasttests""

Actions:
1. hofmannc creates some simple testcases in TestLink (as described in concept for testcase management) focus on crawler load/batch import/release
2. aapetrova implements testcases in Jmeter (Selenium/Jenkins) and learns from this experience for further tests -> focus on simple functionality

Open questions:
1.Testcases should be set in parallel somehow in a clever and reproducable way. How could this be realized with this setting?
2. How many requests/users/items should be set/scaled?
 -> to be discussed later: how to implement user coming from different servers? is this relevant?

**Comment by hofmannc** [comment]
Created 8 test steps in TestLink Testsuite "Load Test".
Test steps are still vague and did not create Testcase-Excel file or excel data table since we are still exploring possibilities. 
Detailed test steps, test cases and data tables are defined later based on the insights we will gain from this exploration.

**Comment by aapetrova** [comment]
**PubMan test case implementation process:**
1. Separate test cases are written as JUnit test case classes
2. Eclipse generates .jar files of every test case
3. A new test plan in JMeter is created
4. A JUnit Request in JMeter is added to test plan and a .jar file is imported as a source
5. JMeter performs test, whereby the number of thread groups (users) will be determined later
6. Results are displayed in a table with a pass/fail status and time of execution in ms

**Comment by hofmannc** [comment]
info for load test values:
- User/get requests
  on a long-term basis we can derive values for amount of users and get requests using data of google analytics account for PubMan.
- Post condition/end criteria of the testcases:  1."system still running" 2. response less than 30 minutes
- We should keep in account the bot activity: they are usually attacking the server within a certain timespan (30 or 60 min) at intervals -> the jmeter timespan setting of the testcases will be relevant 
- how to implement user coming from different servers? which infrastructure will be necessary for this? UPDATE 07.07. -> Info Gudrun: For the first step it will not be necessary using several servers, so we will scope this topic later
- System under test (SUT): qa-pubman

--> for the first trial we start with 4 different user accounts (apetrova, chofmann, test_user1, test_user2) and spreading 15 actions over 20 minutes. 
--> how to integrate get-Requests?

**Comment by aapetrova** [comment]
Solved issues with the batch import/delete/release test cases. 
In the Batch Release test case, no items are removed from the Import Workspace like the post-condition says. Is this the intended functionality? If not, this test case will always result in a failure.

**Comment by hofmannc** [comment]
Within the test case it is correct that no item is deleted, but after the test run the items should be deleted respectively removed from the system (since released items cannot be deleted within the system functionality). 
I changed the wording of the post-condition:
" Organizational info for the Selenium implementation:
Please delete released items from system after the test, so the system won't be too overcrowded."

**Comment by hofmannc** [comment]
conclusion meeting 07.07.16:
AP is busy with the implementation of the "real person scenarios" (PM-93,94,95) in Selenium. ---> seems to be solved by now (see comment above)
It seems to make more sense to implement the crawler scenarios in Jmeter. Will be tested next.

meeting 11.07.:
the exact combination of get requests and "person"-actions is to be found out.

First approach: 
test time span: 20 min
1 attack of get request at intervals: e.g. 6 min attack and 2 min pause/crawler (with a bunch of defined actions inspired by Gudrun's log)
-> to be combined with 3 additional crawler actions 
[-> 10 "person actions"]

input by Mathias Walther:

the qa server is not quite stable at the moment, so the first approach might be a bit too much... the combined crawler actions are ok, but it would make more sense to reduce the "person actions".

---> so we'll start with one batch import and 15 search requests spread over the time span (set in parallel to the crawler's get requests).

-> CH extends the load tests in TestLink: search

**Comment by hofmannc** [comment]
Matthias Walther asked if AP implemented the batch deletion of the discarded items via GUI or PubMan ReST-API - he has a tool for the implementation by REST-API. Maybe this would be easier to implement.

AP and MW will check out the tool on Thursday 14.07.
It is AP's decision which method she is going to follow.

update 14.07.: opened new ticket for this topic: #7 
@aapetrova please report there

**Comment by hofmannc** [comment]
This comment is a summary of one of our meetings of 11.07. and to find out if we are on the same page on the general strategy for the implementation of load tests:

**Background**

a) CH assumes the following process for the implementation for load tests for PubMan:
- CH writes teststeps in TestLink describing the single actions (see Testsuite "Loadtests")
- AP implements the "natural person related" actions in Jmeter as independent (gekapselte, also unabhängig voneinander aufrufbare) testcases
- AP implements the crawler's get requests in Selenium as independent testcases 
- AP then arranges all test cases (coming from Jmeter and Selenium) within the "define timespan" functionality in Jmeter
- This timespan containing the special combination of action is the actual QA-load-testcase

--> Question for AP: Is this assumption realistic and adequate for our process?

b) AP found out that original access logs can be imported to Jmeter and transformed to a testcase.

**Proposal for load test strategy**
Goal of the load tests: Cover as much as realistic load scenarios for PubMan as possible with a minimum effort of test implementation.

To be as flexible and near to a realistic load scenario as possible, I propose a 2 way approach for the test release phases:

a) Cover realistic test scenarios:
Create testscenarios on the basis of selected acces logs provided by developers (t.b.d.: potentially relevant access logs)

b) Cover potential load issues:
Define single actions in Selenium and Jmeter and arrange the potential load pattern in Jmeter. 

-> all testcases are saved in a "Library" and for each release testphase/testplan we can select the relevant testcases.

In each testplan for the release both types of load testcases are involved. The scope of the testplan is defined in kickoff phase of the release and discussed with developers and repository manager.

**Comment by aapetrova** [comment]
This seems like a sensible approach. I'll most likely be done with completing the test case library and creating different example test plans by 15.07.

**Comment by hofmannc** [comment]
ok, thank you - let's discuss results in our preparation meeting on Monday, 18.07. 10:00

**Comment by hofmannc** [comment]
new qa-pubman version is online now , so we try this approach for a load test case:

test time span: 20 min
1 attack of get request at intervals: e.g. 6 min attack and 2 min pause/crawler (with a bunch of defined actions inspired by Gudrun's log)
-> to be combined with 3 additional crawler actions
-> 10 "person actions" (2 imports and 8 search requests)

**Comment by hofmannc** [comment]
sent out concept to team -> next step: define Use Cases for load tests
